`allReduce()` function (#1357)
neo4j/docs-cypher: modules/ROOT/pages/functions/predicate.adoc

======

[[functions-allreduce]]

> [role=label--new-2025.08]

Reviewer: Does this need to be labelled as Cypher 25 anywhere? Or is that handled by the drop-down language version selection?

Author: No, this is for the Cypher 25 Manual, so only Neo4j version required.

== allReduce()

.Details
|===
| *Syntax* 3+| `allReduce(accumulator = initial, stepVariable IN list \| reductionFunction, predicate)`

> | *Description* 3+| Returns true if, during the stepwise evaluation of a value across the elements in a given LIST<ANY>, the accumulated result satisfies a specified predicate at every step, allowing for the early pruning of paths that do not meet the predicate.

Reviewer: ```suggestion
| *Description* 3+| Returns true if, during the stepwise evaluation of a value across the elements in a given `LIST<ANY>`, the accumulated result satisfies a specified predicate at every step. Where that list is a group variable defined in a quantified path pattern, it allows for the early pruning of paths that do not satisfy the predicate.
```

It occurred to me that we may want to give a little more of a hint as to how that pruning hook works.

| `stepVariable` | `ANY` | A variable that holds the value of each element of `list` during iteration.
| `list` | `LIST<ANY>` | The list that is being iterated over.
| `reductionFunction` | `ANY` | An expression whose return value becomes the next value of the `accumulator`.
The return type must match the return type of `initial`.
| `predicate` | `BOOLEAN` | A predicate that is evaluated for each iteration.
It has access to the variable `accumulator`, but not `stepVariable`. 
| *Returns* 3+| `BOOLEAN`
|===

.Considerations
|===
| `allReduce()` differs from most Cypher functions because it iterates over a list, evaluating an expression for each element, rather than returning a result from a single evaluation.
| `allReduce()` combines the functionality of the xref:functions/predicate.adoc#functions-all[`all()`] and xref:functions/list.adoc#functions-reduce[`reduce()`] functions.
| If all evaluations of `predicate` are `true`, `allReduce()` will return `true`.
| If any evaluations of `predicate` are `false`, `allReduce()` will return `false`.
| `allReduce()` returns `true` if `list` is empty because there are no elements to falsify the `predicate`.
|===

.allReduce()
======

> `allReduce()` is designed to optimize path expansions.

Reviewer: Is this meant to be here?

Author: On a second look, no (especially not with your suggested update to the description).

The below query finds `KNOWS` paths with a length of `3` where the cumulative age stays between `60` and `190`.
Paths exceeding this range at any step are pruned.
For example, a path starting with a node whose `age` value is less than `60` is automatically excluded, as is a path with a sequence such as `["Liam Neeson (70)", "Keanu Reeves (58)", "Kathryn Bigelow (71)"]` because its aggregated `age` value exceeds `190`.

.Find aggregated ages within a boundary
// tag::functions_predicate_allreduce_boundary[]
[source, cypher]
----
MATCH (()-[:KNOWS]-(n)){3}
WHERE allReduce(
  acc = 0,
  node IN n | acc + node.age,
  60 < acc < 190
)
RETURN [i IN n | i.name || " (" + toString(i.age) || ")"] AS ageSequence,
       reduce(acc = 0, node IN n | acc + node.age) AS aggregatedAges

> ORDER BY aggregatedAges

Reviewer: The problem with the example is it doesn't show the first person in the chain, due to how `n` inside QPP binds to the right-hand node of each relationship.

One option is to change the query to:
```cypher
MATCH (s) (()-[:KNOWS]-(n)){3}
WHERE allReduce(
 acc = s.age,
 node IN n | acc + node.age,
 acc < 230
)
RETURN [i IN [s] + n | i.name || " (" + toString(i.age) || ")"] AS ageSequence,
      reduce(acc = 0, node IN [s] + n | acc + node.age) AS aggregatedAges
ORDER BY aggregatedAges
```

Another option is to side-step having to bring in that awkward fact about QPP node binding and change the example to find paths showing chains of actors whose knowing is temporally ordered, because there the property is on the relationships:
```cypher
MATCH (s) (()-[:KNOWS]-(n)){3}
WHERE allReduce(
 acc = s.age,
 node IN n | acc + node.age,
 acc < 230
)
RETURN [i IN [s] + n | i.name || " (" + toString(i.age) || ")"] AS ageSequence,
      reduce(acc = 0, node IN [s] + n | acc + node.age) AS aggregatedAges
ORDER BY aggregatedAges
```

Reviewer: Just saw the next example! OK it looks like you have that one covered ... So my first alternative is what I am suggesting now.

----
// end::functions_predicate_allreduce_boundary[]

.Result
[role="queryresult",options="header,footer",cols="2*<m"]
|===
| ageSequence | aggregatedAges

| ["Liam Neeson (70)", "Guy Pearce (55)", "Carrie Anne Moss (55)"] | 180
| ["Liam Neeson (70)", "Keanu Reeves (58)", "Carrie Anne Moss (55)"] | 183
| ["Kathryn Bigelow (71)", "Keanu Reeves (58)", "Carrie Anne Moss (55)"] | 184    

2+d|Rows: 3
|===

> The next query uses `alReduce()` to compare neighboring relationships.

Reviewer: ```suggestion
The next query uses `allReduce()` to compare neighboring relationships.
```

> It finds KNOWS paths with a length of at least `3` where each relationship’s `since` value is greater than the previous one and above `2000`.

Reviewer: ```suggestion
It finds `KNOWS` paths with a length of at least `3` where each relationship’s `since` value is greater than the previous one and above `2000`.
```
